pages/quests/index.tsx

```diff
@@ -8,24 +8,111 @@ import {
+import { useUpdateToken } from '@/hooks/quest/useUpdateToken'
 import { useHasNFT } from '@/hooks/useHasNFT'
+import { useTokenIdOf } from '@/hooks/useTokenIdOf'
+import { useTokenURI } from '@/hooks/useTokenURI'
+import { getContractAddress } from '@/utils/contractAddress'
+import { useIpfsSubmit, KamonToken } from '@/hooks/useIpfsSubmit'
+import { useIpfsGet } from '@/hooks/useIpfsGet'
 
 const Quests: NextPage = () => {
   const { t } = useTranslation('common')
   const mounted = useMounted()
   const { connect, connectors } = useConnect()
   const [metaMask] = connectors
   const { data } = useAccount()
-  const { keyword, inputChange, submit, isSubmitting } = useKeywordSubmit()
+  const { keyword, inputChange, submit, isSubmitting, keywordSubmitSucceeded } = useKeywordSubmit()
   const { hasNFT } = useHasNFT()
+  const { activeChain } = useNetwork()
+  const kamonNFT = getContractAddress({
+    name: 'kamonNFT',
+    chainId: activeChain?.id
+  })
+  const { tokenIdOf } = useTokenIdOf(kamonNFT, data?.address)
+  const { tokenURI } = useTokenURI(kamonNFT, tokenIdOf?.toNumber() || 0)
+  const [tokenId, setTokenId] = useState<BigInt>(BigInt(0))
+  const [tokenJSON, setTokenJSON] = useState<KamonToken>()
+  const [finalTokenUri, setFinalTokenUri] = useState('')
+  const [newTokenImageURI, setNewTokenImageURI] = useState('')
+  const [newTokenJSON, setNewTokenJSON] = useState<KamonToken>()
+  const [updateTxFailed, setUpdateTxFailed] = useState<boolean>()
+
+  const { ipfsSubmit, ipfsSubmitSucceeded, ipfsSubmitIsSubmitting } = useIpfsSubmit()
+  const { ipfsGet, ipfsGetIsSubmitting } = useIpfsGet()
+  const { updateToken, updateTokenSucceeded, updateTokenIsSubmitting } = useUpdateToken(
+    kamonNFT,
+    tokenId,
+    finalTokenUri,
+  )
+  const submittingSomething = () => {
+    return isSubmitting || ipfsSubmitIsSubmitting || ipfsGetIsSubmitting || updateTokenIsSubmitting
+  }
+  useEffect(() => {
+    if (tokenURI) {
+      const fetchData = async () => {
+        const pinataRequest = await fetch(tokenURI.toString())
+        const responseJson = await pinataRequest.json()
+        setTokenJSON(responseJson)
+      }
+      fetchData()
+    }
+  }, [tokenURI])
+
+  useEffect(() => {
+    if (keywordSubmitSucceeded && !ipfsSubmitIsSubmitting
+      && ipfsSubmitSucceeded == undefined
+    ) {
+      if (data?.address == undefined || tokenJSON == undefined) return
+      const userAddress: string = data?.address
+      const ipfsSubmitWrapper = async () => {
+        const ipfsSubmitRet = await ipfsSubmit(tokenJSON, userAddress)
+        if (ipfsSubmitRet !== 'error' && ipfsSubmitRet !== 'same_points') {
+          const tempFinalTokenUri = await ipfsSubmitRet
+          setFinalTokenUri(tempFinalTokenUri)
+        }
+      }
+      setUpdateTxFailed(false)
+      ipfsSubmitWrapper()
+    }
+  }, [keywordSubmitSucceeded, data?.address, tokenJSON, ipfsSubmit, ipfsSubmitIsSubmitting, ipfsSubmitSucceeded])
```

Comment:
can you separate useEffect hooks to a different file?
and I don't think all of the dependency is necessary. 
what exactly do you wanna do here ?

Reply:
@SamuraiT This might be the only thing pending.

At this point, we have this flow:
- User clicks on "Submit Keyword"
- User clicks on "Update NFT", which calls `updateTokenMetadataWrapper()`
- `updateTokenMetadataWrapper` changes state triggering a hook
- The hook changes state again, triggering a second hook

The last two hooks are just functions that run once each sequentially.

I think I should make these into regular functions and skip useEffect(). I can just call each function at the end of the previous one! This will skip all triggering conditions and dependencies. Plus, I might be able to remove state variables that aren't needed for rendering.

Does this make sense?

PS: I did move the toast calls to these functions as per our meeting, so maybe we should keep the functions on this file if they are clear and short enough?